### ISSUE TITLE
Access log has multiple copies of the same entry

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2102]
|<img alt="Cyrus Daboo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Cyrus Daboo reported|
|-|-|
|Trac ID|trac#2102|
|Type|defect|
|Created|2006-09-18 18:09:03Z|

Attached is a fix for access logging in Twisted whereby multiple entries get written into the log file - one for each call to locateChild during request processing.

Attachments:

* [accesslog.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a09/a09c727b76c26eecd467ae2a97db6a11240e2a7d/fce202e40d03d4acedf0cfd3e39b334b414241e6.patch) (1106 bytes) - added by Cyrus Daboo on 2006-09-18 18:10:35Z - Patch to fix problem.
* [test_log.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a09/a09c727b76c26eecd467ae2a97db6a11240e2a7d/a7f9ffb7acf73d619f4b81b3da8e019e95fae83c.patch) (2700 bytes) - added by jknight on 2006-12-27 18:13:05Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2102 2102
type__defect defect
reporter__Cyrus_Daboo Cyrus Daboo
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web2 web2
keywords__ 
time__1158602943000000 1158602943000000
changetime__1171013050000000 1171013050000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

This needs a unit test.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_2"></a>|@jyknight commented|
|-|-|

This sounds like it's probably an API misunderstanding. LogWrapperResource should only be wrapped around the outermost resource. However, I did notice that it's not tested that there aren't duplicate log messages, so here's a patch to test_log.py to assert that things work right.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_4"></a>|@jerub set owner to @jyknight|
|-|-|

The following pyflakes warnings exist in test_log:
```
twisted/web2/test/test_log.py:1: 'unittest' imported but unused
twisted/web2/test/test_log.py:2: 'iweb' imported but unused
twisted/web2/test/test_log.py:6: 'FakeFile' imported but unused
```

There's no copyright header at the top of the file. I believe there should be a copyright header.

There is no docstring on assertLogged. Please add a docstring to this method. SetDateWrapperResource and BufferingLogObserver should have brief docstrings.

Thankyou for updating these tests. :)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve removed owner|
|-|-|

Ready to review in web2-test-log-2102.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_6"></a>|@dreid set owner to @therve|
|-|-|

The following lines in test_log.py don't affect the outcome of the tests, so appear to be completely unecessary
```
        self.resrc.addSlash = True
        self.resrc.responseHeaders = {'Date': 0.0}
```

Other than that this branch is ready to merge.  Thank you for your time therve.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set status to `closed`|
|-|-|

(In [19587]) Merge web2-test-log-2102

Author: jknight, therve
Reviewer: jerub, dreid
Fixes [#2102](https://github.com/twisted/twisted/issues/2102)

Update web2 log tests to be sure there were no duplicate messages. Also
make some cleanups in the process.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
